src/encrypted_dir.rs

```diff
@@ -0,0 +1,552 @@
+    DeleteError, LockError, OpenDirectoryError, OpenReadError, OpenWriteError,
+};
+use tantivy::directory::Directory;
+use tantivy::directory::WatchHandle;
+use tantivy::directory::{
+    AntiCallToken, DirectoryLock, Lock, ReadOnlySource, TerminatingWrite, WatchCallback, WritePtr,
+};
+
+use zeroize::Zeroizing;
+
+use crate::encrypted_stream::{AesReader, AesWriter};
+
+/// KeyBuffer type that makes sure that the buffer is zeroed out before being
+/// dropped.
+type KeyBuffer = Zeroizing<Vec<u8>>;
+
+/// Key derivation result type for our initial key derivation. Consists of a
+/// tuple containing a encryption key, a MAC key, and a random salt.
+type InitialKeyDerivationResult = (KeyBuffer, KeyBuffer, Vec<u8>);
+
+/// Key derivation result for our subsequent key derivations. The salt will be
+/// read from our key file and we will re-derive our encryption and MAC keys.
+type KeyDerivationResult = (KeyBuffer, KeyBuffer);
+
+// The constants here are chosen to be similar to the constants for the Matrix
+// key export format[1].
+// [1] https://matrix.org/docs/spec/client_server/r0.5.0#key-exports
+const KEYFILE: &str = "seshat-index.key";
+// 16 byte random salt.
+const SALT_SIZE: usize = 16;
+// 16 byte random IV for the AES-CTR mode.
+const IV_SIZE: usize = 16;
+// 32 byte or 256 bit encryption keys.
+const KEY_SIZE: usize = 32;
+// 32 byte message authentication code since HMAC-SHA256 is used.
+const MAC_LENGTH: usize = 32;
+// 1 byte for the store version.
+const VERSION: u8 = 1;
+
+#[cfg(test)]
+// Tests don't need to protect against brute force attacks.
+const PBKDF_COUNT: u32 = 10;
+
+#[cfg(not(test))]
+// This is quite a bit lower than the spec since the encrypted key and index
+// will not be public. An attacker would need to get hold of the encrypted index,
+// its key and only then move on to bruteforcing the key. Even if the attacker
+// decrypts the index he wouldn't get to the events itself, only to the index
+// of them.
+const PBKDF_COUNT: u32 = 10_000;
+
+#[derive(Clone, Debug)]
+/// A Directory implementation that wraps a MmapDirectory and adds AES based
+/// encryption to the file read/write operations.
+pub struct EncryptedMmapDirectory {
+    path: PathBuf,
+    mmap_dir: tantivy::directory::MmapDirectory,
+    encryption_key: KeyBuffer,
+    mac_key: KeyBuffer,
+}
+
+impl EncryptedMmapDirectory {
+    /// Open a encrypted mmap directory. If the directory is empty a new
+    /// directory key will be generated and encrypted with the given passphrase.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - The path where the directory should reside in.
+    /// * `passphrase` - The passphrase.
+    ///
+    /// Returns an error if the path does not exist, if it is not a directory or
+    /// if there was an error when trying to decrypt the directory key e.g. the
+    /// given passphrase was incorrect.
+    pub fn open<P: AsRef<Path>>(path: P, passphrase: &str) -> Result<Self, OpenDirectoryError> {
```

Comment:
Can you document how the encryption works?

Reply:
Sure, but probably not in this user facing documentation. Not that this struct is public anyways.

Reply:
I added the documentation to the struct level documentation: https://github.com/matrix-org/seshat/pull/15/commits/d625012b9c1404ad84fb820defbf999152b64ed9.